packages/@aws-cdk/aws-cloudwatch/lib/metric-util.ts

```diff
@@ -0,0 +1,308 @@
+        // Nothing
+      },
+      withExpression(expr) {
+        for (const [id, subMetric] of Object.entries(expr.usingMetrics)) {
+          const existing = seen.get(id);
+          if (existing && metricKey(existing) !== metricKey(subMetric)) {
+            throw new Error(`Same id ('${id}') used for two metrics in the expression: '${subMetric}' and '${existing}'. Rename one.`);
+          }
+          seen.set(id, subMetric);
+          visit(subMetric);
+        }
+      }
+    });
+  }
+}
+
+const METRICKEY_SYMBOL = Symbol('@aws-cdk/aws-cloudwatch.MetricKey');
+
+/**
+ * Return a unique string representation for this metric.
+ *
+ * Can be used to determine as a hash key to determine if 2 Metric objects
+ * represent the same metric. Excludes rendering properties.
+ */
+export function metricKey(metric: IMetric): string {
+  // Cache on the object itself. This is safe because Metric objects are immutable.
+  if (metric.hasOwnProperty(METRICKEY_SYMBOL)) {
+    return (metric as any)[METRICKEY_SYMBOL];
+  }
+
+  const parts = new Array<string>();
+
+  const conf = metric.toMetricConfig();
+  if (conf.mathExpression) {
+    parts.push(conf.mathExpression.expression);
+    for (const id of Object.keys(conf.mathExpression.usingMetrics).sort()) {
+      parts.push(id);
+      parts.push(metricKey(conf.mathExpression.usingMetrics[id]));
+    }
+  }
+  if (conf.metricStat) {
+    parts.push(conf.metricStat.namespace);
+    parts.push(conf.metricStat.metricName);
+    for (const dim of conf.metricStat.dimensions || []) {
+      parts.push(dim.name);
+      parts.push(dim.value);
+    }
+    if (conf.metricStat.statistic) {
+      parts.push(conf.metricStat.statistic);
+    }
+    if (conf.metricStat.period) {
+      parts.push(`${conf.metricStat.period.toSeconds()}`);
+    }
+    if (conf.metricStat.region) {
+      parts.push(conf.metricStat.region);
+    }
+    if (conf.metricStat.account) {
+      parts.push(conf.metricStat.account);
+    }
+  }
+
+  const ret = parts.join('|');
+  Object.defineProperty(metric, METRICKEY_SYMBOL, { value: ret });
+  return ret;
+}
+
+/**
+ * Return the period of a metric
+ *
+ * For a stat metric, return the immediate period. For an expression metric,
+ * return the longest period of its constituent metrics (because the math expression
+ * will only emit a data point if all of its parts emit a data point).
+ *
+ * Formally it should be the LCM of the constituent periods, but the max is simpler,
+ * periods are likely to be multiples of one another anyway, and this is only used
+ * for display purposes.
+ */
+export function metricPeriod(metric: IMetric): Duration {
+  return dispatchMetric(metric, {
+    withStat(stat) {
+      return stat.period;
+    },
+    withExpression(expr) {
+      const metrs = Object.values(expr.usingMetrics);
+      let maxPeriod = Duration.seconds(0);
+      for (const metr of metrs) {
+        const p = metricPeriod(metr);
+        if (p.toSeconds() > maxPeriod.toSeconds()) {
```

Comment:
Now that period is inherited from top-level expression, It's **not** expected to maximize anything here.
In fact if they're different that means that period inheritance isn't working right?